// webservice/webservice.js
var request = require('request')
var clone = require('clone')

var Webservice = {
	host : null, 
	port : null, 
	init : function(host, port) {
		this.host = host
		this.port = port

		return clone(this)
	},
	get : function(path, callback) {
		request.get('http://' + this.host + ':' + this.port + '/' + path, function (err, response, body) {
			if (!err && response.statusCode == 200) {
				callback(body)
			}
			if (err) {
				callback(require('../error/error').dispatch(4, err))
			}
		})
	},
	write : function(path, data, callback) {
		return request.post('http://' + this.host + ':' + this.port + path, { form : data }, function (err, response, body) {
			if (!err && response.statusCode == 200) {
				return callback(body)
			}
			if (err) {
				return callback(require('../error/error').dispatch(4, err))
			}
		})
	}
}

var webservice

module.exports = {
	setWebService : function(host, port) {
		webservice = Webservice.init(host, port)
	},
	getWebService : function() {
		return webservice
	}
}
